lua/wire/zvm/zvm_opcodes.lua

```diff
@@ -542,14 +542,39 @@ end
 ZVM.OpcodeTable[78] = function(self)  --MCOPY
   self:Dyn_EmitForceRegisterLocal("ESI")
   self:Dyn_EmitForceRegisterLocal("EDI")
+  self:Dyn_Emit("$L VAL")
   self:Dyn_Emit("for i = 1,math.Clamp($1,0,8192) do")
-    self:Dyn_Emit("$L VAL")
     self:Dyn_Emit("VAL = VM:ReadCell(ESI)")
     self:Dyn_EmitInterruptCheck()
     self:Dyn_Emit("VM:WriteCell(EDI,VAL)")
     self:Dyn_EmitInterruptCheck()
+    self:Dyn_BeginQuotaOnlyCode()
+        self:Dyn_Emit("VM.MCOPYWrapUpCount = math.Clamp($1,0,8192)-i")
+    self:Dyn_EndQuotaOnlyCode()
     self:Dyn_Emit("EDI = EDI + 1")
     self:Dyn_Emit("ESI = ESI + 1")
+    self:Dyn_StartQuotaInterrupt()
+        self:Dyn_Emit("$L VAL")
+        self:Dyn_Emit("for i = 1, math.min(4096,VM.MCOPYWrapUpCount) do")
+          self:Dyn_Emit("VAL = VM:ReadCell(ESI)")
+          self:Dyn_EmitInterruptCheck()
+          self:Dyn_Emit("VM:WriteCell(EDI,VAL)")
+          self:Dyn_EmitInterruptCheck()
+          self:Dyn_Emit("EDI = EDI + 1")
+          self:Dyn_Emit("ESI = ESI + 1")
+          self:Dyn_BeginQuotaOnlyCode()
+            self:Dyn_EmitState()
+            self:Dyn_Emit("VM.MCOPYWrapUpCount = VM.MCOPYWrapUpCount - i")
+            self:Dyn_Emit("return")
+          self:Dyn_EndQuotaOnlyCode()
+        self:Dyn_Emit("end")
+        self:Dyn_Emit("if VM.MCOPYWrapUpCount <= 4096 then")
+          self:Dyn_Emit("VM.MCOPYWrapUpCount = nil")
+          self:Dyn_Emit("VM.QuotaOverrunFunc = nil")
+          self:Dyn_Emit("else")
+          self:Dyn_Emit("VM.MCOPYWrapUpCount = VM.MCOPYWrapUpCount - 4096")
+        self:Dyn_Emit("end")
+    self:Dyn_EndQuotaInterrupt()
```

Comment:
> It already does that? It kind of looked to me like it was throwing an interrupt that cancels the instruction entirely, but I guess that's wrong.

### Dyn_QuotaOnlyCode details
QuotaOnlyCode will wrap the emits between the start and end with code that will allow it to check if it's going to exit, similar to the Interrupt, but this will allow you to set up things ahead of time that will be needed by the quota interrupt function

On line 551-553 it sets up the variable VM.MCOPYWrapUpCount
On line 565-569 it will emit the state of the CPU, subtract the indexes copied during this execution from the stored variable, and then exit early without changing the quota interrupt func

### Dyn_QuotaInterrupt details

If quota is enabled, and the TIMER goes past the QUOTA, which is defined as TIMER+CPU Cycles, it will place a function containing all of this code from the dyn_emits between StartQuotaInterrupt and EndQuotaInterrupt in VM.QuotaOverrunFunc, it will also add the size of the current instruction to the instruction pointer, so when it wakes up and starts running VM:Step again it will start from the next instruction, it will also end a block like it's a branching(JMP/CALL/etc.) instruction, so precompile should end on "quota" instructions.

When the CPU sees that VM:QuotaOverrunFunc is not nil, it will run it instead of the regular VM:Step, until VM:QuotaOverrunFunc is nil again.

The function wrapped inside behaves like MCOPY, the diffs are:
Capped at 4096 per copy rather than 8192
Tracks variable VM.MCOPYWrapUpCount to access and store the amount of indexes left to copy.
If it's completed the amount of indexes, it will remove itself as function, and remove the variable it used to track, next tick will return to execution at the next instruction.

Reply:
I intended this review to actually point to more than just this section of the file but idk how to correct that